p9/messages.go

```diff
@@ -1999,6 +2000,192 @@ func (r *rusymlink) String() string {
+const (
+	lockOK LockStatus = iota
+	lockBlocked
+	lockError
+	lockGrace
+)
+
+// ELockError is the overarching error for Tlock.
+// The 9P2000.L lock errors are basically "something failed",
+// i.e., not conformant with how 9P2000 is supposed to work.
+var ELockError = syscall.Errno(lockError)
+
+// The Linux v9fs client implements the fcntl(F_SETLKW) (blocking)
+// lock request by calling lock with P9_LOCK_FLAGS_BLOCK set. If the
+// response is P9_LOCK_BLOCKED, it retries the lock request in an
+// interruptible loop until status is no longer P9_LOCK_BLOCKED.
+
+// The Linux v9fs client translates BSD advisory locks (flock) to
+// whole-file POSIX record locks. v9fs does not implement mandatory
+// locks and will return ENOLCK if use is attempted.
+
+// Because of POSIX record lock inheritance and upgrade properties,
+// pass-through servers must be implemented carefully.
+
+// tlock is a Tlock message
+type tlock struct {
+	// fid is the fid to lock.
+	fid fid
+
+	Type   LockType // Type of lock: F_RDLCK, F_WRLCK, F_UNLCK */
+	Flags  uint32   // flags, not whence, docs are wrong.
+	Start  uint64   // Starting offset for lock
+	Length uint64   // Number of bytes to lock
+	PID    int32    // PID of process blocking our lock (F_GETLK only)
+
+	// "client_id is an additional mechanism for uniquely
+	// identifying the lock requester and is set to the nodename
+	// by the Linux v9fs client."
+	// https://github.com/chaos/diod/blob/master/protocol.md#lock---acquire-or-release-a-posix-record-lock
+	Client string // Client id -- but technically can be anything.
+}
+
+// decode implements encoder.decode.
+func (t *tlock) decode(b *buffer) {
+	t.fid = b.ReadFID()
+	t.Type = LockType(b.Read8())
+	t.Flags = b.Read32()
+	t.Start = b.Read64()
+	t.Length = b.Read64()
+	t.PID = int32(b.Read32())
+	t.Client = b.ReadString()
+}
+
+// encode implements encoder.encode.
+func (t *tlock) encode(b *buffer) {
+	b.WriteFID(t.fid)
+	b.Write8(uint8(t.Type))
+	b.Write32(t.Flags)
+	b.Write64(t.Start)
+	b.Write64(t.Length)
+	b.Write32(uint32(t.PID))
+	b.WriteString(t.Client)
+}
+
+// typ implements message.typ.
+func (*tlock) typ() msgType {
+	return msgTlock
+}
+
+// String implements fmt.Stringer.
+func (t *tlock) String() string {
+	return fmt.Sprintf("Tlock{Type: %s, Flags: %#x, Start: %d, Length: %d, PID: %d, Client: %s}", t.Type.String(), t.Flags, t.Start, t.Length, t.PID, t.Client)
+}
+
+// rversion is a version response.
+type rlock struct {
+	Status uint8
+}
+
+// decode implements encoder.decode.
+func (r *rlock) decode(b *buffer) {
+	r.Status = b.Read8()
+}
+
+// encode implements encoder.encode.
+func (r *rlock) encode(b *buffer) {
+	b.Write8(r.Status)
+}
+
+// typ implements message.typ.
+func (*rlock) typ() msgType {
+	return msgRlock
+}
+
+// String implements fmt.Stringer.
+func (r *rlock) String() string {
+	return fmt.Sprintf("Rlock{Status: %d}", r.Status)
+}
+
+// Let's wait until we need this? POSIX locks over a network make 0 sense.
```

Comment:
without adding to the registry, things would still fail, e.g.:
```
2022/08/23 19:32:22 unknown error: invalid message type: 52
go: RLock /var/app/go.mod: input/output error
```
this is running `go` from a Docker image over `cpu`

so I added this to the registry:
```
  msgDotLRegistry.register(msgTlock, func() message { return &tlock{} })
  msgDotLRegistry.register(msgRlock, func() message { return &rlock{} })
```

and I get:
```
 ---- Rlock{Status: 0}
 ---- Rlock{Status: 2}
go: RLock /var/app/go.mod: no locks available
```

a bit further! (debug prints added to `encode` to gain some insight)

Reply:
so I just override the status with `0` (still printing the value _before_, just ignore here), aaaaaaand....
```
docker run -e CPU=192.168.22.209 -e CMD='/bin/bash -c "/go/bin/go run .; ls -l /tmp; cat /tmp/cpu*"' -t go-rv-cpu
cpu command: /bin/bash -c "/go/bin/go run .; ls -l /tmp; cat /tmp/cpu*"
cpu namespace: /usr:/bin:/sbin:/lib:/lib64:/var:/etc:/run:/home:/go:/app:/tmp
 ---- Rlock{Status: 0}
----- got 9p version: 9P2000.L
 ---- Rlock{Status: 2}
 ---- Rlock{Status: 2}
# hello
/go/pkg/tool/linux_riscv64/link: mapping output file failed: invalid argument
go: failed to remove work dir: unlinkat /tmp/go-build755732045/b001/_gomod_.go: function not implemented
```

this is `go run .` on a hello world that doesn't know the world :wave:

Reply:
Something was weird though because it took quite some time and the `init` went up from few percents to around 80% CPU usage for all the time Go compiled, though the `go` process itself only took 3-4% of the CPU.

I checked using the `top` from openSUSE's rootfs via another Docker image.